pmd-cli/src/main/java/net/sourceforge/pmd/cli/commands/internal/PmdCommand.java

```diff
@@ -0,0 +1,484 @@
+                    + "If the file doesn't exist, it will be created on the first run. The file will be overwritten on each run "
+                    + "with the most up-to-date rule violations.")
+    public void setCacheLocation(final Path cacheLocation) {
+        this.cacheLocation = cacheLocation;
+    }
+
+    @Option(names = "--no-cache", description = "Explicitly disable incremental analysis. The '-cache' option is ignored if this switch is present in the command line.")
+    public void setNoCache(final boolean noCache) {
+        this.noCache = noCache;
+    }
+
+    @Option(names = { "--threads", "-t" }, description = "Sets the number of threads used by PMD.",
+            defaultValue = "1")
+    public void setThreads(final int threads) {
+        if (threads < 0) {
+            throw new ParameterException(spec.commandLine(), "Thread count should be a positive number or zero, found " + threads + " instead.");
+        }
+        
+        this.threads = threads;
+    }
+
+    @Option(names = "--no-progress", description = "Disables progress bar indicator of live analysis progress.")
+    public void setNoProgressBar(final boolean noProgressBar) {
+        this.noProgressBar = noProgressBar;
+    }
+
+    /**
+     * Converts these parameters into a configuration.
+     *
+     * @return A new PMDConfiguration corresponding to these parameters
+     *
+     * @throws ParameterException if the parameters are inconsistent or incomplete
+     */
+    public PMDConfiguration toConfiguration() {
+        final PMDConfiguration configuration = new PMDConfiguration();
+        configuration.setInputPaths(inputPaths.stream().map(Path::toString).collect(Collectors.toList()));
+        configuration.setInputFilePath(fileListPath != null ? fileListPath.toString() : null);
+        configuration.setIgnoreFilePath(ignoreListPath != null ? ignoreListPath.toString() : null);
+        configuration.setInputUri(uri != null ? uri.toString() : null);
+        configuration.setReportFormat(format);
+        configuration.setDebug(debug);
+        configuration.setSourceEncoding(encoding.name());
+        configuration.setMinimumPriority(minimumPriority);
+        configuration.setReportFile(reportFile != null ? reportFile.toString() : null);
+        configuration.setReportProperties(properties);
+        configuration.setReportShortNames(shortnames);
+        configuration.setRuleSets(rulesets);
+        configuration.setRuleSetFactoryCompatibilityEnabled(!this.noRuleSetCompatibility);
+        configuration.setShowSuppressedViolations(showSuppressed);
+        configuration.setSuppressMarker(suppressMarker);
+        configuration.setThreads(threads);
+        configuration.setFailOnViolation(failOnViolation);
+        configuration.setAnalysisCacheLocation(cacheLocation != null ? cacheLocation.toString() : null);
+        configuration.setIgnoreIncrementalAnalysis(noCache);
+        configuration.setProgressBar(!noProgressBar);
```

Comment:
@oowekyala I'm considering removing this from the configuration and `PmdAnalysis`, as it's exclusively intended for CLI usage… however, since instantiating the listener requires to know the final number of files to analyze, this requires some additional changes in `PmdAnalysis`. Therefore, I'll refrain from doing it in this PR and leave it for a follow up. WDTY?

Reply:
You're right that this should be refactored. Maybe we should have an own "ProgressListener" interface in pmd-core which the cli module could implement?

Reply:
I'll look into that after this PR is closed.